examples/Reporting/ParallelReportDownload.cs

```diff
@@ -0,0 +1,242 @@
+        /// <param name="client">The Google Ads client.</param>
+        /// <param name="customerIds">The Google Ads customer Id.</param>
+        /// <param name="loginCustomerId">Optional login customer ID if your access to the CIDs
+        ///     is via a manager account.</param>
+        public void Run(GoogleAdsClient client, string customerIds, string loginCustomerId = null)
+        {
+            // If a manager ID is supplied, update the login credentials.
+            if (loginCustomerId != null)
+            {
+                client.Config.LoginCustomerId = loginCustomerId;
+            }
+
+            // Get the GoogleAdsService. A single client can be shared by all threads.
+            GoogleAdsServiceClient googleAdsService = client.GetService(
+                Services.V3.GoogleAdsService);
+
+            List<long> ids = customerIds.Split(',').Select(long.Parse).ToList();
+
+            try
+            {
+                // Begin downloading reports and block program termination until complete.
+                Task task = RunDownloadParallelAsync(googleAdsService, ids);
+                task.Wait();
+            }
+            catch (GoogleAdsException e)
+            {
+                Console.WriteLine("Failure:");
+                Console.WriteLine($"Message: {e.Message}");
+                Console.WriteLine($"Failure: {e.Failure}");
+                Console.WriteLine($"Request ID: {e.RequestId}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Initiate all download requests, wait for their completion, and report the results.
+        /// </summary>
+        /// <param name="googleAdsService">The Google Ads service.</param>
+        /// <param name="customerIds">The list of customer IDs from which to request data.</param>
+        /// <returns>The asynchronous operation.</returns>
+        private async Task RunDownloadParallelAsync(
+            GoogleAdsServiceClient googleAdsService, List<long> customerIds)
+        {
+            // List of all requests to ensure that we wait for the reports to complete on all
+            // customer IDs before proceeding.
+            List<Task<SearchStreamStream>> tasks = new List<Task<SearchStreamStream>>();
+
+            // Collection of downloaded responses.
+            ConcurrentBag<ReportDownload> responses = new ConcurrentBag<ReportDownload>();
+
+            // Helpful identifier for active tasks.
+            int reportId = 1;
+
+            // IMPORTANT: You should avoid hitting the same customer ID in parallel. There are rate
+            // limits at the customer ID level which are much stricter than limits at the developer
+            // token level. Hitting these limits frequently enough will significantly reduce
+            // throughput as the client library will automatically retry with exponential back-off
+            // before failing the request.
+            foreach (string query in GAQL_QUERY_STRINGS)
```

Comment:
Perhaps there should be some kind of indicator on the reportdownload object to help identify what type of report this is.

Reply:
Agreed, I was thinking about that, but thought the most clutter-free way was to have a numeric identifier rather than, say, re-printing the query.  I'd like to capture the Request ID, but couldn't see a way to extract it from the stream or response.  Thoughts?

Reply:
Reworked to show friendly query name from dictionary key.